### ISSUE TITLE
Example for passive (push-based) pipeline usage

### ISSUE THREAD

**Issue by hendrik-cliqz** [thomasmoelhave/tpie#205]
I am trying to use pipelining in a way that I want to push items into it myself, this is described as 'passive' in the documentation, still all examples use an 'active' stage in the beginning. 

In code it looks e.g. like that, first initializing a passive sorter and a pipeline based on it:

    tpie::tpie_init();
    tpie::get_memory_manager().set_limit(50*1024*1024);
    tpie::pipelining::passive_sorter<int> sorter;
    auto input = sorter.input();
    tpie::pipelining::pipeline p = pullFromSort(sorter.output()) | tpie::pipelining::printf_ints();

Now I try to use input, by initializing it and pusing items:

    auto x = input.construct();
    x.begin();
    x.push(20);
    x.push(40);
    x.end();
    tpie::tpie_finish();

Unfortunately this ends in a segfault. I do not see what I am doing wrong. Can you help me? It would be good to have a working example for such a case.

For completeness the definition of pullFromSort:

    template <typename source_t>
    class PullFromSort {
    public:
      template <typename dest_t>
      class type : public tpie::pipelining::node {
      public:
        typedef int item_type;

        type(dest_t dest, source_t src)
          : dest(std::move(dest))
          , puller(src.construct())
        {
          add_push_destination(dest);
          add_pull_source(puller);
          set_name("PullFromSort");
        }

        virtual void go() override {
          while (puller.can_pull()) {
            int number = puller.pull();
            dest.push(number);
          }
        }

      private:
        dest_t dest;
        typedef typename source_t::constructed_type puller_t;
        puller_t puller;
      };
    };

    template <typename source_t>
    inline tpie::pipelining::pipe_begin<tpie::pipelining::tempfactory<PullFromSort<source_t>, source_t> > pullFromSort(source_t source) {
      return {std::move(source)};
    }

**Comment by antialize** [comment]
Hello can you perhaps detail what you are trying to do and why?  I see
4 solutions to your problem listed in order of usefulness.

1) In the wast majority of cases you do not need a "passive" pipeline
as you have described, you can almost always create a pipe to go
before a sorter that produces the items instead.

2) If all you want is to sort items, perhaps you do not need a
pipeline. You can just use a sorter object, you can supply items to it
and consume items from it. See here
https://github.com/thomasmoelhave/tpie/blob/master/tpie/sort.h#L66 for
an example.

3) You can use a (somewhat badly named) subpipeline, though this
causes a virtual call per element (hardly noticeable in the grand
scheme of things), we do not currently have good documentation for
this,  but the interface is fairly stable

namespace tp=tpie::pipelining;
tp::subpipeline<int> sp = tp::sort() | tp::printf_ints();
sp.begin(get_available_memory());
sp.push(20);
sp.push(40);
sp.end();

4) You can do what you are trying to do in the example however the
interfaces needed to do this are not stable and do change from time to
time, also they are not well documented as they are meant for internal
use only.

Some of the many things that are wrong in your example is that no
memory have been assigned to the pipenodes, that no phase graph has
been computed and that prepare, propagate, begin, go, and end is not
called on all nodes in the pipeline in their proper order. (I am
probably forgetting a lot of things here). This is somewhat hard to do
right, however it can obviously be done (as we do it).  We have not
documented how to do this as you are not meant to do it, however you
can see how it is done in
https://github.com/thomasmoelhave/tpie/blob/master/tpie/pipelining/runtime.h
and https://github.com/thomasmoelhave/tpie/blob/master/tpie/pipelining/runtime.cpp

/Jakob

**Comment by hendrik-cliqz** [comment]
Thanks for the answer:

To explain my usecase: Right now I am already [using](https://github.com/cliqz-oss/keyvi/blob/master/keyvi/src/cpp/dictionary/dictionary_compiler.h) the external sorter (I started using TPIE before pipelining went into master) and I am happy with it. My API basically allows the client to push items ( Add() ) and finalize ( Compile() ), after that the client can persist it to a file or stream. So basically what you describe as solution 2.

My workflow in a nutshell looks like this:

    key_value -> parse_value -> value_dedup_and_index -> sorter(with parallel sort) -> key_index(single_threaded)

So why do I want to switch to pipelining: the stage parse_value which involves some string parsing and compression is the main bottleneck, therefore I want to run it in parallel:

    key_value -> parallel(parse_value) -> value_dedup_and_index -> sorter(with parallel sort) -> key_index(single_threaded)

The pipelining framework - as I understood - has some highly efficient implementation to do this. That's the reason why I wanted to try this approach. As this is the beginning of the pipeline I could as well implement my own way, but why re-invent the wheel.

As to your last comments: I totally agree with you that using internal API's is not the way to go. 1 better solution would be slightly modified version of the push_input_iterator implemented using the official API's: Instead of an iterator taking a queue and pop elements until it receives an end signal.

Appreciate your comments on this idea.

**Comment by antialize** [comment]
Note that the sort and serialization_sort in pipelining is just a wrapper
around the merger_sorter and the serialization_sorter, so your current code
will not see a speedup by switching to pipelining.  Note however that
serialization sorting is currently quite slow, and might be made much
faster.

Currently we perform a quicksort on the serialized elements in memory, so
elements are serialized once, but deserialized fully for each comparison in
the sorter.

We could speed this up quite a lot, however it would require some thinking
and some more code. In essens if we do not store a serialized version of
the item when doing internal sort, we need some way of letting the user
provide a method to tell us how much memory a specific items takes up of
memory, if we had such a method, then we could just std::move the item into
a std::vector and sort it using our parallel sort.  We can try to get the
maintainer to do this, but it is probably not on the top of the stack
currently.

/Jakob

**Comment by hendrik-cliqz** [comment]
I want to speed up value parsing which involves in my case JSON parsing (rapidjson), msgpack serialization (msgpack-c) and compression(snappy/zlib), that is independent of the sorting. This is done right before sorting and as said could be done in parallel.

As for sorting speed: I think uses 20% of the runtime, it is not the major bottleneck. By indexing the values before sort this got much faster: Earlier I sorted string pairs: key and value, now I sort a string and an id, which is much smaller in size, less copying and therefore faster.

**Comment by hendrik-cliqz** [comment]
Closing this issue as it seems out of scope. There are other ways to implement my requirement. 

Thanks for the support!
